### PR TITLE
Relax "no-null-undefined-union" rule.

### PR DESCRIPTION
**Pull request by nrathi**
#### PR checklist

- [x] Addresses an existing issue: fixes https://github.com/palantir/tslint/issues/4617
- [ ] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### CHANGELOG.md entry:
[rule-change] Relax `no-null-undefined-union` rule.

### CODE REVIEW

**Comment by VincentLanglet** [comment]
Hi @nrathi 

Can you add a test for
```
type Text = string | null

interface Props {
  title?: Text
}
interface SuperProps {
  title?: Text | number
}

function someFunc(bar?: Text, foo?: Text | number) {}
```

I don't expect error from the rule, if you do, there should be at least an option to disable these cases.

**Comment by nrathi** [comment]
@VincentLanglet added, no errors!

File: test/rules/no-null-undefined-union/test.ts.lint [hunk 1]
```diff
@@ -52,4 +58,18 @@ function someFunc() {
 }
 ~ [0]
 
+const z = testFunc();
+
+type Text = string | null
+
+interface TextInterface {
+  foo?: Text
```
Comment (adidahiya) [inline test/rules/no-null-undefined-union/test.ts.lint]: 1. indentation is inconsistent on this line (use 4 spaces)
2. I don't agree with this change... this is just syntactic sugar for `Text | undefined`, which is in turn sugar for `string | null | undefined`. Why should we allow this? What's the point of the rule? How do you explain to devs which style is allowed and which isn't?

Comment (VincentLanglet) [reply test/rules/no-null-undefined-union/test.ts.lint]: Plus `?` is just not syntactic sugar for Text | undefined.
```
interface A {
 a: number | undefined
}
interface B {
 b?: number
}

const a: A = {} // $error
const b: B = {} // no error
```

When you declare props in react, all I saw
```
interface Props {
  requiredValue: SomeType
  optionalValue?: SomeType
}

<Component requiredValue={...}  />
```

I tried
```
interface Props {
  value: SomeType | undefined
}

<Component /> // $error property value is missing
```

Plus actually the rule create conflict with the `non-null-keyword` in the `tslint:all` config. With the precedent props, you have to use the component this way
```
interface Props {
  value: SomeTypeCouldBeNull
}

<Component value={null} /> // $error prefer use undefined
```

Every user can disable the rules he does not want, but at least the proposal `tslint:all` should not have conflict.

Comment (adidahiya) [reply test/rules/no-null-undefined-union/test.ts.lint]: ok, that's fair, thanks for the examples. after looking at this again, the rule makes more sense now. I think we should update the documentation to also explain this, though. we should mention that interface fields and function params with `?` are expected to be exempt from the rule.

**Pull request merged**